File: cpp/cpp_100-999/687.cc
// C++
// leetcode 687
// https://leetcode.cn/problems/longest-univalue-path/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
	// 返回值为，包含结点 p 的单向长度最长长度
	// res 为 global 的可往两个子树延伸的最长长度
	int dfs(TreeNode* p, int &res) {
		// 空结点直接返回 0
		if (p == nullptr)
			return 0;

		// l, r存储左右子树的答案
		int l = dfs(p->left, res), r = dfs(p->right, res);

		// 若左子树与当前结点值相同则 +1，否则归零
		if (p->left && p->left->val == p->val)
			l++;
		else
			l = 0;

		// 右子树同理
		if (p->right && p->right->val == p->val)
			r++;
		else
			r = 0;

		// 结算一次 res
		res = max(res, l + r);
		
		return max(l, r);
	}

    int longestUnivaluePath(TreeNode* root) {
    	int res = 0;
    	dfs(root, res);
    	return res;
    }
};

File: Go/Go_100-999/687.go
// Go
// leetcode 687
// https://leetcode.cn/problems/longest-univalue-path/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func longestUnivaluePath(root *TreeNode) int {
	res := 0
	dfs(root, &res)
	return res
}

func dfs(p *TreeNode, res *int) int {
	if p == nil {
		return 0
	}

	l, r := dfs(p.Left, res), dfs(p.Right, res)
	if p.Left != nil && p.Left.Val == p.Val {
		l++
	} else {
		l = 0
	}
	if p.Right != nil && p.Right.Val == p.Val {
		r++
	} else {
		r = 0
	}

	if *res < l+r {
		*res = l + r
	}

	if l > r {
		return l
	}
	return r
}